cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+				zinfo->io_ack_waiting = 1;
+				pthread_cond_wait(&zinfo->io_ack_cond,
+				    &zinfo->complete_queue_mutex);
+
+				zinfo->io_ack_waiting = 0;
+				if ((zinfo->state == ZVOL_INFO_STATE_OFFLINE) ||
+				    (zinfo->conn_closed == true)) {
+					(void) pthread_mutex_unlock(
+					    &zinfo->complete_queue_mutex);
+					goto exit;
+				}
+			}
+		} while (STAILQ_EMPTY(&zinfo->complete_queue));
+
+		zio_cmd = STAILQ_FIRST(&zinfo->complete_queue);
+		STAILQ_REMOVE_HEAD(&zinfo->complete_queue, cmd_link);
+		(void) pthread_mutex_unlock(&zinfo->complete_queue_mutex);
+
+		ASSERT(zio_cmd->conn == fd);
+		ZREPL_LOG("ACK for op:%d with seq-id %ld\n",
+		    zio_cmd->hdr.opcode, zio_cmd->hdr.io_seq);
+
+		rc = uzfs_zvol_socket_write(zio_cmd->conn,
+		    (char *)&zio_cmd->hdr, sizeof (zio_cmd->hdr));
+		if (rc == -1) {
+			ZREPL_ERRLOG("socket write err :%d\n", errno);
+			zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+			goto exit;
+		}
+
+		switch (zio_cmd->hdr.opcode) {
+			case ZVOL_OPCODE_HANDSHAKE:
+			case ZVOL_OPCODE_WRITE:
+			case ZVOL_OPCODE_SYNC:
+				zinfo->write_req_ack_cnt++;
+				/* Send handsake ack */
+				break;
+			case ZVOL_OPCODE_READ:
+				printf("ACK for op:%d with seq-id %ld\n",
+				    zio_cmd->hdr.opcode, zio_cmd->hdr.io_seq);
+				/* Send data read from disk */
+				rc = uzfs_zvol_socket_write(zio_cmd->conn,
+				    zio_cmd->buf,
+				    (sizeof (char) * zio_cmd->hdr.len));
+				if (rc == -1) {
+					ZREPL_ERRLOG("socket write err :%d\n",
+					    errno);
+					ASSERT(0);
+					goto exit;
+				}
+				zinfo->read_req_ack_cnt++;
+				break;
+
+			default:
+				VERIFY(!"Should be a valid opcode");
+				break;
+		}
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+exit:
+	close(fd);
+	while (!STAILQ_EMPTY(&zinfo->complete_queue)) {
+		zio_cmd = STAILQ_FIRST(&zinfo->complete_queue);
+		STAILQ_REMOVE_HEAD(&zinfo->complete_queue, cmd_link);
+		zio_cmd_free(&zio_cmd, zio_cmd->hdr.opcode);
+	}
+	uzfs_zinfo_drop_refcnt(zinfo, false);
+
+	printf("uzfs_zvol_io_ack_sender thread exiting\n");
+	zk_thread_exit();
+}
+
+static void
+uzfs_zrepl_open_log(void)
+{
+	openlog("zrepl", LOG_PID, LOG_LOCAL7);
+}
+
+static void
+uzfs_zrepl_close_log(void)
+{
+	closelog();
+}
+
+static void
+uzfs_zrepl_walk_pool_directory(void)
+{
+	spa_t	*spa = NULL;
+	spa_t	*sp = NULL;
+	int	rc = 0;
+
+	mutex_enter(&spa_namespace_lock);
+	while ((spa = spa_next(spa)) != NULL) {
+		(void) printf(">>>>>>>\t%s\n", spa_name(spa));
+		rc = spa_open(spa_name(spa), &sp, spa);
+		if (rc == 0) {
+			spa_close(spa, spa);
+		}
+	}
+	mutex_exit(&spa_namespace_lock);
+}
+
+/*
+ * Main function for replica.
+ */
+int
+main(void)
+{
+	int 		rc;
+	kthread_t 	*conn_accpt_thrd;
+	kthread_t 	*uzfs_mgmt_thread;
+
+
+	pthread_t slf = pthread_self();
+	snprintf(tinfo, sizeof (tinfo), "m#%d.%d",
+	    (int)(((uint64_t *)slf)[0]), getpid());
+
+	zfs_arc_max = (512 << 20);
+	zfs_arc_min = (256 << 20);
+	printf("zarcmax: %lu zarcmin:%lu\n", zfs_arc_max, zfs_arc_min);
```

Comment:
shouldn't this be set based on real memory size obtained through linux standard interfaces?  In some rare cases 512MB can be too high while in most of the cases it will be probably too low.

Reply:
Jan to be candid, I have thought abt these values. I have inherited this code from one of file in uzfs. Will see what would be optimal value that we can place here. As of now I am adding TODO comments here.

Reply:
👍